Fix #6562 last written bytebuffer

gregw opened the pull request:
Fixes #6562 the last written bytebuffer calculation.
Also fixed an associated issue with unnecessary flush of an empty when last calculation already signalled last.

gregw committed:
Fix #6562 last written bytebuffer

Fixes #6562 the last written bytebuffer calculation.
Also fixed an associated issue with unnecessary flush of an empty when last calculation already signalled last.

The code coverage is not complete, so more tests are needed for this use case. Also strange that `write(ByteBuffer)` does not appear to every commence aggregation?

gregw committed:
Fix #6562 last written bytebuffer

Removed the last flush of an empty buffer as was no path to that code.

gregw commented:
@lorban @lachlan-roberts In an attempt to get 100% code coverage, I could not find a path to get to the final last emptybuffer write. So I have removed it and I think it is good.  But give me extra good review on that deletion.

lorban commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java:
```diff
@@ -896,10 +896,6 @@ public void write(byte[] b, int off, int len) throws IOException
             }
-            else if (last)
-            {
-                channelWrite(BufferUtil.EMPTY_BUFFER, true);
```
Serving a 0 length `byte[]` goes through this branch. This is why `GzipContentLengthTest.testServletTypeLengthStreamWrite()` fails for scenario `empty.txt`.

gregw replied:
Ah that's how it can be called!!! Restored it and added test cases in `HttpOutputTest` so it is now covered!

lorban commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java:
```diff
@@ -973,8 +969,6 @@ public void write(ByteBuffer buffer) throws IOException
                 // write any remaining content in the buffer directly
                 if (len > 0)
                     channelWrite(buffer, last);
-                else if (last)
```
Serving a 0 length `byte[]` goes through this branch. This is why `GzipContentLengthTest.testHttpOutputWrite()` fails for scenario `empty.txt`.

lorban reviewed (changes requested):
You should probably add some tests to `HttpOutputTest` to assert that writing 0 length buffers does work.

gregw committed:
Fix #6562 last written bytebuffer

Restored the last flush of an empty buffer as is needed for 0 length empty write.
